okio/src/commonMain/kotlin/okio/FileHandle.kt

```diff
@@ -0,0 +1,219 @@
+ * limitations under the License.
+ */
+package okio
+
+/**
+ * An open file for reading and writing; using either streaming and random access.
+ *
+ * Use [read] and [write] to perform one-off random-access reads and writes. Use [source], [sink],
+ * and [appendingSink] for streaming reads and writes.
+ *
+ * File handles must be closed when they are no longer needed. It is an error to read, write, or
+ * create streams after a file handle is closed. The operating system resources held by a file
+ * handle will be released once the file handle **and** all of its streams are closed.
+ *
+ * Although this class offers both reading and writing APIs, file handle instances may be
+ * read-only or write-only. For example, a handle to a file on a read-only file system will throw an
+ * exception if a write is attempted.
+ *
+ * File handles may be used by multiple threads concurrently. But the individual sources and sinks
+ * produced by a file handle are not safe for concurrent use.
+ */
+@ExperimentalFileSystem
+abstract class FileHandle : Closeable {
+  /**
+   * True once the file handle is closed. Resources should be released with [closeInternal] once
+   * this is true and [openStreamCount] is 0.
+   */
+  private var closed = false
+
+  /**
+   * Reference count of the number of open sources and sinks on this file handle. Resources should
+   * be released with [closeInternal] once this is 0 and [closed] is true.
+   */
+  private var openStreamCount = 0
+
+  /**
+   * Removes at least 1, and up to [byteCount] bytes from this and appends them to [sink]. Returns
+   * the number of bytes read, or -1 if this file is exhausted.
+   */
+  @Throws(IOException::class)
+  abstract fun read(offset: Long = 0L, sink: Buffer, byteCount: Long): Long
+
+  /**
+   * Returns the total number of bytes in the file. This will change if the file size changes.
+   */
+  @Throws(IOException::class)
+  abstract fun size(): Long
+
+  /** Removes [byteCount] bytes from [source] and writes them to this at [offset]. */
+  @Throws(IOException::class)
+  abstract fun write(offset: Long = 0L, source: Buffer, byteCount: Long)
+
+  /** Pushes all buffered bytes to their final destination. */
+  @Throws(IOException::class)
+  abstract fun flush()
+
+  /**
+   * Returns a source that reads from this starting at [offset]. The returned source must be closed
+   * when it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun source(offset: Long = 0L): Source {
+    synchronized(this) {
+      check(!closed) { "closed" }
+      openStreamCount++
+    }
+    return FileHandleSource(this, offset)
+  }
+
+  /**
+   * Returns the position of [source] in the file. The argument [source] must be either a source
+   * produced by this file handle, or a [BufferedSource] that directly wraps such a source. If the
+   * parameter is a [BufferedSource], it adjusts for buffered bytes.
+   */
+  @Throws(IOException::class)
+  fun position(source: Source): Long {
```

Comment:
What's the benefit of this API over exposing `FileSystemSource` as a public type with a public `position` property? This could get awkward if doing more than buffering the source/sink, for example, hashing to validate a checksum.

Reply:
Yeah, you’re absolutely right. I definitely wrote that out and tried to see how it’d play out. The biggest drawback of it is how it interacts with `BufferedSource`.

If we introduce a new type then `BufferedSource` either won’t implement the interface, or other `BufferedSource` implementations will implement the interface but not be able to do anything reasonable in this method.

The use case I’m imagining for this is passing a BufferedSource to a method to read some data, then checking the new position.